File: src/rasterize.rs
use crate::Coordinate;
use crate::Stroke;

pub fn world_to_view(coordinate: &Coordinate) -> (usize, usize) {
    let dimension = 9;
    let half = (dimension as f64 - 1.0) / 2.0;
    (
        ((coordinate.x * half + half).round() as usize).min(dimension - 1),
        ((coordinate.y * half + half).round() as usize).min(dimension - 1),
    )
}

pub fn rasterize(strokes: &Vec<Stroke>, dimension: usize) -> Vec<Vec<u8>> {
    let mut array = vec![vec![0u8; dimension]; dimension];

    for stroke in strokes {
        rasterize_stroke(stroke, dimension, &mut array);
    }
    array
}

pub fn rasterize_stroke(stroke: &Stroke, dimension: usize, array: &mut Vec<Vec<u8>>) {
    let (x0, y0) = world_to_view(&stroke.start);
    let (x1, y1) = world_to_view(&stroke.end);
    array[y0][x0] = 1;
    array[y1][x1] = 1;

    if x0 == x1 {
        for y in std::cmp::min(y0, y1)..=std::cmp::min(std::cmp::max(y0, y1), dimension - 1) {
            array[y][x0] = 1;
        }
    } else if y0 == y1 {
        for x in std::cmp::min(x0, x1)..=std::cmp::min(std::cmp::max(x0, x1), dimension - 1) {
            array[y0][x] = 1;
        }
    } else {
        let slope = (x1 as f64 - x0 as f64) / (y1 as f64 - y0 as f64);
        for y in std::cmp::min(y0, y1)..=std::cmp::min(std::cmp::max(y0, y1), dimension - 1) {
            let x =
                ((slope * (y as f64 - y0 as f64) + x0 as f64).floor() as usize).min(dimension - 1);
            array[y][x] = 1;
        }
    }
}

File: glyphs_generator/rasterize.py
from typing import List, Tuple
import numpy as np

from .data import Glyph


def world_to_view(x: float, y: float, glyph_size: int) -> Tuple[int, int]:
    return int(np.round(x * (glyph_size - 1) / 2 + (glyph_size - 1) / 2)), int(
        np.round(y * (glyph_size - 1) / 2 + (glyph_size - 1) / 2)
    )


def rasterize(glyph: Glyph, glyph_size: int) -> np.ndarray:
    array = np.zeros(shape=(glyph_size, glyph_size))

    for stroke in glyph.strokes:
        x0, y0 = world_to_view(x=stroke.x0, y=stroke.y0, glyph_size=glyph_size)
        x1, y1 = world_to_view(x=stroke.x1, y=stroke.y1, glyph_size=glyph_size)
        array[y0, x0] = 1
        array[y1, x1] = 1
        if x0 == x1:
            for y in range(min(y0, y1), min(max(y0, y1) + 1, glyph_size)):
                array[y, x0] = 1
        elif y0 == y1:
            for x in range(min(x0, x1), min(max(x0, x1) + 1, glyph_size)):
                array[y0, x] = 1
        else:
            slope = (x1 - x0) / (y1 - y0)
            for y in range(min(y0, y1), min(max(y0, y1) + 1, glyph_size)):
                x = int(np.floor(slope * (y - y0) + x0))
                array[y, x] = 1
    return array


def rasterize_all(
    glyphs: List[Glyph], glyph_size: int = 5, margin: int = 1, nb_glyphs_per_row: int = 6, sort: bool = False
) -> np.ndarray:
    if not sort:
        nb_glyphs = len(glyphs)
        nb_glyphs_per_columns = int(np.ceil(nb_glyphs / nb_glyphs_per_row))
        tensor = np.zeros(
            shape=(
                (glyph_size + margin) * nb_glyphs_per_columns,
                (glyph_size + margin) * nb_glyphs_per_row,
            )
        )
        for k, glyph in enumerate(glyphs):
            j = k % nb_glyphs_per_row
            i = k // nb_glyphs_per_row
            tensor[
                (glyph_size + margin) * i : (glyph_size + margin) * i + glyph_size,
                (glyph_size + margin) * j : (glyph_size + margin) * j + glyph_size,
            ] = rasterize(glyph=glyph, glyph_size=glyph_size)
        return tensor
    else:
        alphabet = {}
        for glyph in glyphs:
            nb_strokes = len(glyph.strokes)
            if (nb_strokes - 1) not in alphabet:
                alphabet[nb_strokes - 1] = []
            alphabet[nb_strokes - 1].append(glyph)
        largest_number_of_glyphs_per_order = np.max(np.array([len(glyphs) for glyphs in alphabet.values()]))
        strokes = alphabet[len(alphabet) - 1][0].strokes
        tensor = np.zeros(
            shape=(
                (glyph_size + margin) * len(strokes),
                (glyph_size + margin) * largest_number_of_glyphs_per_order,
            )
        )
        for order, glyphs in alphabet.items():
            for i, glyph in enumerate(glyphs):
                tensor[
                    (glyph_size + margin) * order : (glyph_size + margin) * order + glyph_size,
                    (glyph_size + margin) * i : (glyph_size + margin) * i + glyph_size,
                ] = rasterize(glyph=glyph, glyph_size=glyph_size)
        return tensor
